商品購入機能実装

kobayashi-fumiya2 opened the pull request:
# What
商品購入機能実装のため

# Why
商品を購入するため

・必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる動画
https://gyazo.com/1426dec387f0896cd96d4cba40f7c485

・入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画
https://gyazo.com/e220e38452d8675b063726053e595eda

・ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/8b887adb70d62b28b45caa69bbcd551f

・ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する動画
https://gyazo.com/e27be85cd3308c37b16bda88d367e4b3

・ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画
https://gyazo.com/5cdbc827ca5545b74a73cfd9de25a36c

・売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品一覧機能実装時に未実装であった場合）、売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/71d9cf78ad49e73bd5e6411a94bfacb0

・ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/6c61d9c03a1acc33656864fff8a1eee3

・ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
 https://gyazo.com/e8ba3d19c9f90819428c29f4b8cae420

・テスト結果の画像
https://gyazo.com/24fc7d8cba52098e7f023a830520f7d9

Commit:
商品購入機能（クレジット以外）

Commit:
フォームオブジェクト(order_delivery)の単体テストコード

Commit:
token生成・サーバーサイド送信

Commit:
決済機能の実装

Commit:
トークンに対するバリデーション

Commit:
鍵情報を環境変数として設定

Commit:
tokenの単体テストコード

Commit:
ルボコップ

mihot117 commented on app/controllers/items_controller.rb:
```diff
@@ -21,6 +21,7 @@ def create
   end
 
   def show
+    @delivery = Delivery.new
```
```ruby
    @delivery = Delivery.new
```

こちらの変数ですが、コードで確認したところ
show.html.erbのファイル内で使用されていないようです。

不要なコードであれば、削除をお願いいたします。
必要ということであれば、用途をメモに残すか、show.html.erbで実際にご使用ください！

kobayashi-fumiya2 replied:
不要のため削除・修正致しました。

mihot117 commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,39 @@
+    )
+  end
+
+  def pay_item
+    Payjp.api_key = ENV['PAYJP_SECRET_KEY']
+    Payjp.api_key = 'sk_test_b99bc4c21e31a2352076f62b'
+    Payjp::Charge.create(
```
```ruby
    Payjp.api_key = ENV['PAYJP_SECRET_KEY']
    Payjp.api_key = 'sk_test_b99bc4c21e31a2352076f62b'
```

環境変数設定のコードが2行ございますが、
以下の達成要件に則って、不要な方を削除しておきましょう。

```
- 秘密鍵などは、値の見える状態でGitHubにpushしないこと。環境変数化してからcommitすること。
```

kobayashi-fumiya2 replied:
修正致しました。

mihot117 commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,39 @@
+class OrdersController < ApplicationController
+  before_action :authenticate_user!
+
+  def index
+    @item = Item.find(params[:item_id])
+    @order_delivery = OrderDelivery.new
+    redirect_to root_path if @item.user_id == current_user.id || @item.order
+  end
+
+  def create
+    @item = Item.find(params[:item_id])
+    @order_delivery = OrderDelivery.new(order_delivery_params)
+    if @order_delivery.valid?
+      pay_item
+      @order_delivery.save
+      redirect_to root_path
```
```ruby
@item = Item.find(params[:item_id])
```

index, createアクション内で上記の記述が重複していますので、リファクタリングをお願いいたします。

kobayashi-fumiya2 replied:
修正致しました。

mihot117 commented on app/models/order_delivery.rb:
```diff
@@ -0,0 +1,17 @@
+  include ActiveModel::Model
+  attr_accessor :postcode, :prefecture_id, :city, :block, :building, :phone_number, :order_id, :item_id, :user_id, :token
+
+  validates :postcode,      presence: true, format: { with: /\A\d{3}-\d{4}\z/ }
+  validates :prefecture_id, presence: true, numericality: { other_than: 1, message: "can't be blank" }
+  validates :city,          presence: true
+  validates :block,         presence: true
+  validates :phone_number,  presence: true, format: { with: /\A\d{10,11}\z/ }
+  validates :token, presence: true
```
5〜10行目のバリデーションにて`presence`の記述が重複していますので、リファクタリングしておきましょう！

`with_options`を使用することで簡単にまとめることが可能です。
詳細は以のカリキュラムをご確認ください。
```
最終課題補足カリキュラム
〜1つのフォームから複数のテーブルに情報を保存するアプリを作成しよう〜
```

kobayashi-fumiya2 replied:
修正致しました。加えて、itemモデルとuserモデルのリファクタリングも行いました。

mihot117 commented on spec/models/order_delivery_spec.rb:
```diff
@@ -0,0 +1,65 @@
+        @order_delivery.valid?
+        expect(@order_delivery.errors.full_messages).to include("City can't be blank")
+      end
+      it 'blockが空では保存できない' do
+        @order_delivery.block = ''
+        @order_delivery.valid?
+        expect(@order_delivery.errors.full_messages).to include("Block can't be blank")
+      end
+      it 'blockが空では保存できない' do
+        @order_delivery.block = ''
+        @order_delivery.valid?
+        expect(@order_delivery.errors.full_messages).to include("Block can't be blank")
+      end
```
内容が重複しているようにお見受けいたしますので
片方だけ削除いただいて問題ないかと存じます！

kobayashi-fumiya2 replied:
重複しておりました。修正致しました。

mihot117 commented on spec/models/order_delivery_spec.rb:
```diff
@@ -0,0 +1,65 @@
+        @order_delivery.valid?
+        expect(@order_delivery.errors.full_messages).to include("Block can't be blank")
+      end
+      it 'phone_numberが空では保存できない' do
+        @order_delivery.phone_number = ''
+        @order_delivery.valid?
+        expect(@order_delivery.errors.full_messages).to include("Phone number can't be blank", 'Phone number is invalid')
```
```ruby
include("Phone number can't be blank", 'Phone number is invalid')
```

エラー文が２つ出力された場合は`今回のテストに適切なエラー文のみを残して`片方は削除いただいて問題ございません。

テストコードでは、テストの内容に応じて特定のエラー文が出ることを確認する必要があるため、片方が不要となります。

kobayashi-fumiya2 replied:
修正致しました。

mihot117 commented on spec/models/order_delivery_spec.rb:
```diff
@@ -0,0 +1,65 @@
+        @order_delivery.valid?
+        expect(@order_delivery.errors.full_messages).to include("Phone number can't be blank", 'Phone number is invalid')
+      end
+      it 'phone_numberが10桁以上11桁以内の半角数値以外は保存できない' do
+        @order_delivery.phone_number = '123-456-789'
+        @order_delivery.valid?
+        expect(@order_delivery.errors.full_messages).to include('Phone number is invalid')
```
電話番号に関しましては、下記のテストコードの記述もお願いいたします。

バリデーションが適用されているかを確かめるためです。

```
・全角数字だと登録できないこと
・9桁以下だと登録できないこと
・12桁以上だと登録できないこと
```

kobayashi-fumiya2 replied:
修正致しました。

mihot117 commented on spec/models/order_delivery_spec.rb:
```diff
@@ -0,0 +1,65 @@
+        @order_delivery.token = ''
+        @order_delivery.valid?
+        expect(@order_delivery.errors.full_messages).to include("Token can't be blank")
+      end
```
テストコードに下記の追記をお願いいたします。

```
■購入者情報
・userが紐付いていなければ購入できない
■商品情報
・itemが紐付いていなければ購入できない
```

user, itemの情報が必須であることを確かめるためです。

kobayashi-fumiya2 replied:
修正致しました。

mihot117 reviewed (commented):
レビュー依頼ありがとうございます。
コメントいたしましたので、ご確認いただけますと幸いです。

修正後、再度レビュー依頼フォームのご提出ください。
また、今回はテストコードの修正が含まれていますので
次回ご提出いただく際は最新のテスト成功Gyazoを頂戴してもよろしいでしょうか？

ご対応のほど、よろしくお願いいたします。

Commit:
不要なコード削除

Commit:
環境変数にしていない秘密鍵の削除

Commit:
コントローラーの重複箇所のリファクタリング

Commit:
Formオブジェクト(order_delivery)のリファクタリング

presence: trueをwith_optionsでまとめた

Commit:
itemモデルのリファクタリング

presence: true、numericality: { other_than: 1, message: "can't be blank" }をwith_optionsでまとめた

Commit:
userモデルのリファクタリング

presence: trueをwith_optionsでまとめた

Commit:
テストコードの重複削除

Commit:
テストコード修正

エラー文が２つ出力されたが今回のテストに適切なエラー文のみを残した。

Commit:
テストコード修正

電話番号に関してのバリデーション確認のため

Commit:
order_delivelyモデル、テストコード修正

user,item情報のバリデーション、テストコード漏れのため。

Commit:
ルボコップ

fumiron commented:
レビューありがとうございます！
修正致しました。

下記テスト成功Gyazoになります。
https://gyazo.com/ecbff518bae43dd4bc2224fa4aa6e4b2

mihot117 reviewed (commented):
レビュー依頼ありがとうございます！

修正点を確認いたしました。

**LGTM**とさせていただきます！
実装お疲れ様でした！

こちらのブランチは忘れずにMergeいただき、ローカルのMasterにPullした後に
次の実装のためのブランチを切って次の課題に取り組みましょう！

引き続き、開発頑張っていきましょう！応援しております。
